File: frontend/src/types/site.ts
export interface Zone {
  id: string;
  label: string;
  x: number;
  y: number;
  width: number;
  height: number;
  phase: string;
  phase_progress: number;
  /** Multi-level: which level this zone lives on. Defaults to "L0"
   *  (ground floor) for legacy single-floor projects. */
  level_id?: string;
}

export interface ScheduleEntry {
  zone_id: string;
  phase: string;
  start_day: number;
  end_day: number;
  trades_required: string[];
}

export interface Level {
  id: string;
  name: string;
  elevation_m: number;
  order: number;
  background_image_url?: string | null;
}

export interface ConnectionNode {
  level_id: string;
  x: number;
  y: number;
}

export interface SiteConnection {
  id: string;
  kind: "stair" | "elevator";
  nodes: ConnectionNode[];
}

export interface Road {
  /** Author-stable id; not used by the renderer beyond debug labels. */
  id: string;
  /** Polyline points in site metres. The renderer stamps a strip of
   *  `width_m` along every segment + a half-width disk at every node. */
  points: [number, number][];
  /** Width of the stamped strip in metres. */
  width_m: number;
  /** Which level the road lives on. Defaults to "L0" for legacy
   *  single-floor projects. */
  level_id?: string;
}

export interface Site {
  id: string;
  name: string;
  width: number;
  height: number;
  zones: Zone[];
  current_day: number;
  schedule: ScheduleEntry[];
  /** Phase 2: levels are always present. Single-floor projects expose
   *  exactly one level (L0). */
  levels?: Level[];
  /** Phase 3: vertical-transport graph. Empty for single-floor projects. */
  connections?: SiteConnection[];
  discipline?: string;
  /** Authored walkable corridors (polylines). When empty the renderer
   *  falls back to a default south + west perimeter strip so legacy
   *  documents still draw something sensible. */
  roads?: Road[];
}

File: backend/models/site.py
from pydantic import BaseModel
from enum import Enum

from models.assets import DEFAULT_LEVEL_ID


class Phase(str, Enum):
    EXCAVATION = "excavation"
    # Tiefbau-specific phases (Phase 5 wires the renderer + behaviour).
    SHORING = "shoring"
    PILING = "piling"
    DRAINAGE = "drainage"
    FOUNDATION = "foundation"
    STRUCTURAL = "structural"
    MEP_ROUGHIN = "mep_roughin"
    CLOSEIN = "closein"
    FINISHES = "finishes"
    PAVING = "paving"
    COMPLETE = "complete"


class Discipline(str, Enum):
    """Construction discipline tag on a project. Drives default level
    structure, palette of available phases, and renderer hints."""

    HOCHBAU = "hochbau"
    TIEFBAU = "tiefbau"
    HYBRID = "hybrid"


class Level(BaseModel):
    """A named vertical slice of the site. Workers move within a level
    continuously; cross-level travel goes through a `Connection` (stair
    or elevator). `elevation_m` is the floor's height above grade; can
    be negative for Tiefgarage / UG levels."""

    id: str
    name: str
    elevation_m: float
    order: int
    background_image_url: str | None = None


class Zone(BaseModel):
    id: str
    label: str
    x: float
    y: float
    width: float
    height: float
    phase: Phase
    phase_progress: float
    level_id: str = DEFAULT_LEVEL_ID


class ScheduleEntry(BaseModel):
    zone_id: str
    phase: Phase
    start_day: int
    end_day: int
    trades_required: list[str]


class Road(BaseModel):
    """Authored walkable corridor. A polyline of `(x, y)` points; the
    navmesh + renderer stamp a strip of `width_m` along each segment.

    Roads are the cheapest cells on the navmesh — workers prefer them
    over open ground and over zone interiors. Drawing them is the
    canvas equivalent: a darker strip with a dashed centre line.
    """

    id: str
    points: list[tuple[float, float]]
    width_m: float = 6.0
    level_id: str = "L0"


class Site(BaseModel):
    id: str
    name: str
    width: float
    height: float
    zones: list[Zone]
    current_day: int
    schedule: list[ScheduleEntry] = []
    discipline: Discipline = Discipline.HOCHBAU
    levels: list[Level] = []
    # Authored road / pedestrian-corridor network. When empty the navmesh
    # and renderer fall back to the legacy hardcoded perimeter pattern
    # (south + west strips). Polyline-based so authors can curve roads
    # around obstacles rather than be limited to grid-aligned strips.
    roads: list[Road] = []
